mono/metadata/icall.c

```diff
@@ -1162,36 +1162,38 @@ ves_icall_System_ValueType_Equals (MonoObject *this_obj, MonoObject *that, MonoA
 			continue;
 		if (mono_field_is_deleted (field))
 			continue;
+		guint8 *this_field = (guint8 *) this_obj + field->offset;
+		guint8 *that_field = (guint8 *) that + field->offset;
 		/* FIXME: Add more types */
 		switch (field->type->type) {
 		case MONO_TYPE_U1:
 		case MONO_TYPE_I1:
 		case MONO_TYPE_BOOLEAN:
-			if (*((guint8*)this_obj + field->offset) != *((guint8*)that + field->offset))
+			if (memcmp (this_field, that_field, sizeof (guint8)))
 				return FALSE;
 			break;
 		case MONO_TYPE_U2:
 		case MONO_TYPE_I2:
 		case MONO_TYPE_CHAR:
-			if (*(gint16*)((guint8*)this_obj + field->offset) != *(gint16*)((guint8*)that + field->offset))
+			if (memcmp (this_field, that_field, sizeof (guint16)))
 				return FALSE;
 			break;
 		case MONO_TYPE_U4:
 		case MONO_TYPE_I4:
-			if (*(gint32*)((guint8*)this_obj + field->offset) != *(gint32*)((guint8*)that + field->offset))
+			if (memcmp (this_field, that_field, sizeof (guint32)))
 				return FALSE;
 			break;
 		case MONO_TYPE_U8:
 		case MONO_TYPE_I8:
-			if (*(gint64*)((guint8*)this_obj + field->offset) != *(gint64*)((guint8*)that + field->offset))
+			if (memcmp (this_field, that_field, sizeof (guint64)))
 				return FALSE;
 			break;
 		case MONO_TYPE_R4:
-			if (*(float*)((guint8*)this_obj + field->offset) != *(float*)((guint8*)that + field->offset))
+			if (memcmp (this_field, that_field, sizeof (float)))
```

Comment:
This is going to be wrong for NaNs (and infinities?)

Reply:
shouldn't it be the same representation in memory?

Reply:
`float x = 0.0f / 0.0;   int always_false = (x == x)` a NaN is never equal to itself.

Reply:
🙁  you are right.

Now that means we need to do something like this:
```c
case MONO_TYPE_R4: {
  float left, right;
  memcpy (&left, this_field, sizeof (float));
  memcpy (&right, that_field, sizeof (float));
  if (left != right)
    return FALSE;
  break;
}
```

I don't like it 😕

Reply:
If the only case that poses problem is `NaN`, then we should use`isnan`; it's a `libc` function that is available everywhere.

Reply:
```if (*(uint32_t*)&left) != *(uint32_t*)&right)) return false;``` ? Though it may or may not be faster or work. i.e. Does the compiler notice the hardcoded memcpy size? What about alignment? memcpy will handle that, mine will not. If memcpy isn't optimized, and alignment matters, you can also, like: ``` if (!(this_field & 3) && !(that_field & 3) ...) which will slow it down to maybe speed i up.